README.md
py-jstree
=========

An introduction will be proposed later.

Installation
------------

If you have Python installed and wish to get the package directly from the
[Python Package Index](http://pypi.python.org/pypi/jstree), just run
`pip install jstree` from the command-line.  If you already have a prior
version installed, just run `pip install jstree -U` instead.

Contribute
----------

Please help contribute to this project by going to the
[GitHub Project Repository](https://github.com/grimwm/py-jstree) and doing one
of a few things:

 * send me pull requests through the github interface
 * point me directly to your git repo so I can pull changes
 * send bug reports and feature requests by filing them under the __Issues__ tab at the top

Examples
--------
    >>> import jstree
    >>> node = jstree.Node('a', None)
    >>> print node
    Node({'text': 'a', 'children': MutableDictionaryObject({})})
    >>> print node.jsonData()
    {'text': 'a'}

    >>> import jstree
    >>> node = jstree.Node('a', 1)
    >>> print node
    Node({'text': 'a', 'children': MutableDictionaryObject({}), 'li_attr': DictionyObject({'id': 1}), 'id': 1})
    >>> print node.jsonData()
    {'text': 'a', 'id': 1, 'li_attr': {'id': 1}}

    >>> import jstree
    >>> node = jstree.Node('a', 5, icon="folder", state = {'opened': True})
    >>> print node
    Node({'text': 'a', 'id': 5, 'state': DictionaryObject({'opened': True}), 'children': MutableDictionaryObject({}), 'li_attr': DictionaryObject({'id': 5}), 'icon': 'folder'})
    >>> print node.jsonData()
    {'text': 'a', 'state': {'opened': True}, 'id': 5, 'li_attr': {'id': 5}, 'icon': 'folder'}

    >>> import jstree
    >>> paths = [jstree.Path("editor/2012-07/31/.classpath", 1), jstree.Path("editor/2012-07/31/.project", 2)]
    >>> print jstree.JSTree(paths).pretty()
    /
      editor/
        2012-07/
          31/
            .classpath
            .project

    >>> import jstree
    >>> paths = [jstree.Path("editor/2012-07/31/.classpath", 1), jstree.Path("editor/2012-07/31/.project", 2)]
    >>> t = jstree.JSTree(paths)
    >>> d = t.jsonData()
    >>> print d
    [{'data': 'editor', 'children': [{'data': '2012-07', 'children': [{'data': '31', 'children': [{'data': '.classpath', 'metadata': {'id': 1}}, {'data': '.project', 'metadata': {'id': 2}}]}]}]}]
    >>> print d[0]['children'][0]['children'][0]['children'][1]
    {'data': '.project', 'metadata': {'id': 2}}
    >>> print d[0]['children'][0]['children'][0]['children'][1]['data']
    .project
    >>> print d[0]['children'][0]['children'][0]['children'][1]['metadata']['id']
    2


CHANGELOG.md
Changes
=======

v0.5
====
* Add Python 3 support.

v0.4
====
* Removed the 'id' tag from general data.
* Renamed 'metadata' to 'li_attr'.

v0.3
====
* Enhanced JSTree so it can add IDs to leaf nodes for selection in
  a jsTree.  Although, this may need more work after testing in
  a jsTree application.
* Updated all the doctests to handle the additional burden
  of users being able to pass in arbitrary attributes and
  "path object ids".
* Removed ability to copy JSTrees directly in __init__ since it's
  not actually a use-case right now.  Trying to avoid code smells.

v0.2
====
* Started updating the changelog, finally.
* Simplified the tree data structure by removing the insertion of
  "data" at the top-level.  This means JS clients will have to add
  it themselves when doing full-loads of data off the server, but
  it also means that AJAX-based jsTrees should work properly.  More
  testing on this is required, however.
* Started enhancing the in-code documentation so devs get a better
  sense of how to use the code.
* Hardened jstree.Node to be "as immutable as possible" right now
  without a lot of major code changes.
* Enhanced jstree.Node for advanced usage.  Users can now pass
  additional attributes directly into it if they're managing
  a set of Nodes themselves.

v0.1.2
======
* Fixed 'dictobj' as a requirement of jstree.

v0.1.1
======
* Changed dev status to alpha.  Was incorrectly set to beta.
* Fixed setup.py for users of Python pre-2.7.
* Added Node class to make creation of JSON more manageable.

v0.1
------
* First release.


jstree_test.py
import jstree

import unittest
import doctest

class TestJSTree(unittest.TestCase):
  def setUp(self):
    self.tree = jstree.JSTree([jstree.Path("editor/2012-07/31/.classpath", 1), jstree.Path("editor/2012-07/31/.project", 2)])

  def test_pretty(self):
    output = '/\n  editor/\n    2012-07/\n      31/\n        .classpath\n        .project'
    self.tree.pretty()
    self.assertEqual(output, self.tree.pretty())

def load_tests(loader, tests, pattern):
  suite = unittest.TestSuite()
  suite.addTests(tests)
  suite.addTest(doctest.DocTestSuite(jstree))
  return suite

if '__main__' == __name__:
  unittest.main()


jstree.py
import dictobj
import os
from collections import namedtuple

Path = namedtuple('Path', ('path', 'id'))


class Node(dictobj.DictionaryObject):
  """
  This class exists as a helper to the jsTree.  Its "jsonData" method can
  generate sub-tree JSON without putting the logic directly into the jsTree.

  This data structure is only semi-immutable.  The jsTree uses a directly
  iterative (i.e. no stack is managed) builder pattern to construct a
  tree out of paths.  Therefore, the children are not known in advance, and
  we have to keep the children attribute mutable.
  """

  def __init__(self, path, oid, **kwargs):
    """
    kwargs allows users to pass arbitrary information into a Node that
    will later be output in jsonData().  It allows for more advanced
    configuration than the default path handling that jsTree currently allows.
    For example, users may want to pass "attr" or some other valid jsTree options.

    Example:
      >>> import jstree
      >>> node = jstree.Node('a', None)
      >>> print node
      Node({'text': 'a', 'children': MutableDictionaryObject({})})
      >>> print node.jsonData()
      {'text': 'a'}

      >>> import jstree
      >>> node = jstree.Node('a', 1)
      >>> print node
      Node({'text': 'a', 'children': MutableDictionaryObject({}), 'li_attr': DictionaryObject({'id': 1})})
      >>> print node.jsonData()
      {'text': 'a', 'li_attr': {'id': 1}}

      >>> import jstree
      >>> node = jstree.Node('a', 5, icon="folder", state = {'opened': True})
      >>> print node
      Node({'text': 'a', 'state': DictionaryObject({'opened': True}), 'children': MutableDictionaryObject({}), 'li_attr': DictionaryObject({'id': 5}), 'icon': 'folder'})
      >>> print node.jsonData()
      {'text': 'a', 'state': {'opened': True}, 'li_attr': {'id': 5}, 'icon': 'folder'}
    """
    super(Node, self).__init__()

    children = kwargs.get('children', {})
    if len(filter(lambda key: not isinstance(children[key], Node), children)):
      raise TypeError(
        "One or more children were not instances of '%s'" % Node.__name__)
    if 'children' in kwargs:
      del kwargs['children']
    self._items['children'] = dictobj.MutableDictionaryObject(children)

    if oid is not None:
      li_attr = kwargs.get('li_attr', {})
      li_attr['id'] = oid
      kwargs['li_attr'] = li_attr

    self._items.update(dictobj.DictionaryObject(**kwargs))
    self._items['text'] = path

  def jsonData(self):
    children = [self.children[k].jsonData() for k in sorted(self.children)]
    output = {}
    for k in self._items:
      if 'children' == k:
        continue
      if isinstance(self._items[k], dictobj.DictionaryObject):
        output[k] = self._items[k].asdict()
      else:
        output[k] = self._items[k]
    if len(children):
      output['children'] = children
    return output


class JSTree(dictobj.DictionaryObject):
  """
  An immutable dictionary-like object that converts a list of "paths"
  into a tree structure suitable for jQuery's jsTree.
  """

  def __init__(self, paths, **kwargs):
    """
    Take a list of paths and put them into a tree.  Paths with the same prefix should
    be at the same level in the tree.

    kwargs may be standard jsTree options used at all levels in the tree.  These will be outputted
    in the JSON.

    Example (basic usage):
      >>> import jstree
      >>> paths = [jstree.Path("editor/2012-07/31/.classpath", 1), jstree.Path("editor/2012-07/31/.project", 2)]
      >>> t1 = jstree.JSTree(paths)
      >>> print t1.jsonData()
      [{'text': 'editor', 'children': [{'text': '2012-07', 'children': [{'text': '31', 'children': [{'text': '.classpath', 'li_attr': {'id': 1}}, {'text': '.project', 'li_attr': {'id': 2}}]}]}]}]
    """
    if len(filter(lambda p: not isinstance(p, Path), paths)):
      raise TypeError(
        "All paths must be instances of '%s'" % Path.__name__)

    super(JSTree, self).__init__()

    root = Node('', None, **kwargs)
    for path in sorted(set(paths)):
      curr = root
      subpaths = path.path.split(os.path.sep)
      for i, subpath in enumerate(subpaths):
        if subpath not in curr.children:
          oid = path.id if len(subpaths) - 1 == i else None
          curr.children[subpath] = Node(subpath, oid, **kwargs)
        curr = curr.children[subpath]
    self._items['_root'] = root

  def pretty(self, root=None, depth=0, spacing=2):
    """
    Create a "pretty print" represenation of the tree with customized indentation at each
    level of the tree.

    Example:
    >>> import jstree
    >>> paths = [jstree.Path("editor/2012-07/31/.classpath", 1), jstree.Path("editor/2012-07/31/.project", 2)]
    >>> print jstree.JSTree(paths).pretty()
    /
      editor/
        2012-07/
          31/
            .classpath
            .project
    """
    if root is None:
      root = self._root
    fmt = "%s%s/" if root.children else "%s%s"
    s = fmt % (" " * depth * spacing, root.text)
    for child in root.children:
      child = root.children[child]
      s += "\n%s" % self.pretty(child, depth + 1, spacing)
    return s

  def jsonData(self):
    """
    Returns a copy of the internal tree in a JSON-friendly format,
    ready for consumption by jsTree.  The data is represented as a
    list of dictionaries, each of which are our internal nodes.

    Examples:
    >>> import jstree
    >>> paths = [jstree.Path("editor/2012-07/31/.classpath", 1), jstree.Path("editor/2012-07/31/.project", 2)]
    >>> t = jstree.JSTree(paths)
    >>> d = t.jsonData()
    >>> print d
    [{'text': 'editor', 'children': [{'text': '2012-07', 'children': [{'text': '31', 'children': [{'text': '.classpath', 'li_attr': {'id': 1}}, {'text': '.project', 'li_attr': {'id': 2}}]}]}]}]
    >>> print d[0]['children'][0]['children'][0]['children'][1]
    {'text': '.project', 'li_attr': {'id': 2}}
    >>> print d[0]['children'][0]['children'][0]['children'][1]['text']
    .project
    >>> print d[0]['children'][0]['children'][0]['children'][1]['li_attr']['id']
    2
    """
    return [self._root.children[k].jsonData() for k in sorted(self._root.children)]
